metrics/src/main/java/io/opencensus/metrics/Value.java

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.metrics;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.ExperimentalApi;
+import io.opencensus.common.Function;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * The actual point value for a {@link Point}.
+ *
+ * <p>Currently there are three types of {@link Value}:
+ *
+ * <ul>
+ *   <li>{@link DoubleValue}
+ *   <li>{@link LongValue}
+ *   <li>{@link DistributionValue}
+ * </ul>
+ *
+ * <p>Each {@link Point} contains exactly one of the three {@link Value} types.
+ *
+ * @since 0.16
+ */
+@ExperimentalApi
+@Immutable
+public abstract class Value {
+
+  Value() {}
+
+  /**
+   * Applies the given match function to the underlying data type.
+   *
+   * @since 0.16
+   */
+  public abstract <T> T match(
+      Function<? super DoubleValue, T> p0,
+      Function<? super LongValue, T> p1,
+      Function<? super DistributionValue, T> p2,
+      Function<? super Value, T> defaultFunction);
+
+  /**
+   * A 64-bit double-precision floating-point {@link Value}.
+   *
+   * @since 0.16
```

Comment:
Shouldn't the abstract class name be a prefix?

Reply:
In the [proto](https://github.com/census-instrumentation/opencensus-proto/blob/master/opencensus/proto/stats/metrics/metrics.proto#L150) all the sub-types are suffixed: `int64_value`, `double_value`, `distribution_value`. Do you think we should rename these to `ValueDouble`, `ValueLong` and `ValueDistribution` in Java?

Reply:
Sorry I didn't notice the match() method for [AttributeValue](https://github.com/census-instrumentation/opencensus-java/blob/660e8f375bb483a3eb817940b3aa8534f86da314/api/src/main/java/io/opencensus/trace/AttributeValue.java#L31). I think this approach is better.

Reply:
Make sense to me, and with this approach our APIs are more consistent with the proto definition (where we only have the Distribution as a separate class). WDYT @dinooliva ?

Reply:
I'm good with having match handle Long and Double. Still on the fence about ValueDistribution vs DistributionValue but I prefer to stay consistent with our naming conventions (i.e. ValueDistribution).

On possibility would be to add a static factory method distributionValue() similar to what is done in AttributeValue.

Reply:
SG, I renamed all the subclasses to be Value* (prefixed). Now the subclasses are package-private, and the top-level `Value` class exposes factory methods (`doubleValue()`, `longValue()`, `distributionValue()`) to create each subclass, which is the same to [AttributeValue](https://github.com/census-instrumentation/opencensus-java/blob/660e8f375bb483a3eb817940b3aa8534f86da314/api/src/main/java/io/opencensus/trace/AttributeValue.java#L31).

In addition, I added a top-level `Distribution` class in order to expose its internal stats.